### PR TITLE
Global rate limit support.

### PR DESCRIPTION
**Pull request by konrad-kaminski**

### CODE REVIEW

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/EnvoyDefaultFilters.kt [hunk 1]
```diff
@@ -30,12 +33,17 @@ class EnvoyDefaultFilters(
     private val headerToMetadataHttpFilter = headerToMetadataHttpFilter(defaultHeaderToMetadataConfig)
     private val defaultHeaderToMetadataFilter = { _: Group, _: GlobalSnapshot -> headerToMetadataHttpFilter }
     private val envoyRouterHttpFilter = envoyRouterHttpFilter()
-
     /**
      * Default filters should not be private, user should have an option to pick any filter.
      * Remember: order matters.
      */
     val defaultEnvoyRouterHttpFilter = { _: Group, _: GlobalSnapshot -> envoyRouterHttpFilter }
+    val defaultRateLimitLuaFilter = { group: Group, _: GlobalSnapshot ->
```
Comment (malfoj89) [inline envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/EnvoyDefaultFilters.kt:41]: both can be private

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RateLimitFilterFactoryTest.kt [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+        // then
+        assertThat(filter).isNull()
+    }
+
+    @Test
+    fun `should create ratelimit filter with proper domain and ratelimit service name`() {
+        // given
+        val group = createGroupWithIncomingRateLimits(
+            IncomingRateLimitEndpoint("/hello", rateLimit = "0/s")
```
Comment (malfoj89) [inline envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RateLimitFilterFactoryTest.kt:35]: should it be allowed to setup rate limit to 0?

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/assertions/AwaitAssertions.kt [hunk 3]
```diff
@@ -4,10 +4,10 @@ import org.assertj.core.api.Assertions
 import org.awaitility.Awaitility
 import java.time.Duration
 
-fun <T> untilAsserted(wait: Duration = Duration.ofSeconds(90), fn: () -> (T)): T {
+fun <T> untilAsserted(poll: Duration = Duration.ofMillis(500), wait: Duration = Duration.ofSeconds(90), fn: () -> (T)): T {
```
Comment (malfoj89) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/assertions/AwaitAssertions.kt:7]: I would format this file to have method args in seperate lines

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/consul/ConsulOperations.kt [hunk 4]
```diff
@@ -39,8 +41,9 @@ class ConsulOperations(port: Int) {
         id: String = UUID.randomUUID().toString(),
         name: String,
         registerDefaultCheck: Boolean = false,
-        tags: List<String> = listOf("a")
-    ) = registerService(id, name, extension.container().ipAddress(), extension.container().port(), registerDefaultCheck, tags)
+        tags: List<String> = listOf("a"),
+        beforeRegistration: (NewService) -> Unit = {}
+    ) = registerService(id, name, extension.container().ipAddress(), extension.container().port(), registerDefaultCheck, tags, beforeRegistration)
```
Comment (malfoj89) [inline envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/consul/ConsulOperations.kt:46]: could u format this file?

**Review by malfoj89 — DISMISSED** [review]

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataValidator.kt [hunk 5]
```diff
@@ -41,10 +41,15 @@ class ServiceNameNotProvidedException : NodeMetadataValidationException(
...
 
+class RateLimitIncorrectValidationException(rateLimit: String?) : NodeMetadataValidationException(
+    "Rate limit value: $rateLimit is incorrect."
+)
+
 class NodeMetadataValidator(
     val properties: SnapshotProperties
 ) : DiscoveryServerCallbacks {
     companion object {
+        private val RATE_LIMIT_PATTERN = Regex("^(?:0|[1-9][0-9]*)\\/[s,m,h]$")
```
[the remark this answers is not in the document]
Comment (konrad-kaminski) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataValidator.kt]: Fixed.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataValidator.kt [hunk 6]
```diff
@@ -135,6 +141,22 @@ class NodeMetadataValidator(
         }
     }
 
+    private fun validateIncomingRateLimitEndpoints(metadata: NodeMetadata) {
```
[the remark this answers is not in the document]
Comment (konrad-kaminski) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/groups/NodeMetadataValidator.kt:144]: Done.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt [hunk 7]
```diff
@@ -344,5 +345,10 @@ class JwtFilterProperties {
     var providers = mapOf<ProviderName, OAuthProvider>()
 }
 
+class RateLimitProperties {
```
[the remark this answers is not in the document]
Comment (konrad-kaminski) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/SnapshotProperties.kt]: Fixed.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/EnvoyDefaultFilters.kt (hunk 1 above)
Comment (konrad-kaminski) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/EnvoyDefaultFilters.kt:41]: Nope. See the comment above the declarations.

File: envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RateLimitFilterFactoryTest.kt (hunk 2 above)
Comment (konrad-kaminski) [reply envoy-control-core/src/test/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/RateLimitFilterFactoryTest.kt:35]: Yes. It basically means it's banned.

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/assertions/AwaitAssertions.kt (hunk 3 above)
Comment (konrad-kaminski) [reply envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/assertions/AwaitAssertions.kt]: Fixed.

File: envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/consul/ConsulOperations.kt (hunk 4 above)
Comment (konrad-kaminski) [reply envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/config/consul/ConsulOperations.kt]: Fixed.

File: envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/EnvoyDefaultFilters.kt (hunk 1 above)
Comment (malfoj89) [reply envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/resource/listeners/filters/EnvoyDefaultFilters.kt:41]: cool i am blind apparently :D

**Review by malfoj89 — APPROVED** [review]

**Pull request closed without merging**